docs/cadence/language/references.mdx

```diff
@@ -268,11 +268,103 @@ resource I {
+Inside the body of function with entitlement-mapped access,
+the name of the entitlement mapping may be used as a stand-in for the output entitlements of the map.
 
-## Reference validity
+## Field and Index Access
+
+References to container types (structs/resources, dictionaries and arrays) can be used to access (read/write) fields
+or elements of the container.
+
+When a field/index is read through a reference, it wil return:
+- A reference, if the field / value at index is also container-typed.
+- Or, the concrete value, if the value is a primitive type.
+
+For example, consider the below `Collection` resource which has two fields: one (id) is String-typed, 
+and the other (ownedNFTs) is dictionary-typed.
+
+```cadence
+resource Collection {
+
+    // Primitive-typed field
+    access(all) var id: String
+
+    // Dictionary typed field
+    access(all) var ownedNFTs: @{UInt64: NFT}
+}
+```
+
+Thus,
+
+```cadence
+var collectionRef: &Collection = ...
+
+// `collectionRef.ownedNFTs` would return a reference of type `&{UInt64: NFT}`.
+var ownedNFTsRef: &{UInt64: NFT} = collectionRef.ownedNFTs
+
+// Whereas, `collectionRef.id` would return the value, since it is a primitive type.
+var id: String = collectionRef.id
+```
+
+Similarly, accessing an element of an array/dictionary will return a reference.
+
+```cadence
+// Index-access to an array reference would return a reference to the element.
+var resourceArrayRef: &[AnyResource] = ...
+var elementRef: &AnyResource = collectionArrayRef[2]
+
+// Whereas, if the array is of primitive type, it will return the concrete value.
+var intArrayRef: &[Int] = ...
+var element: Int = intArrayRef[2]
+```
+
+```cadence
+// Index-access to a dictionary reference would return a refe rence to the value.
+var resourceDictionaryRef: &{String: AnyResource} = ...
+var valueRef: &AnyResource? = resourceDictionaryRef["two"]
+
+// Whereas, if the dictionary values are of primitive type, it will return the concrete value.
+var intDictionaryRef: &{String: Int} = ...
+var value: Int? = intDictionaryRef["two"]
+```
+
+It is also important to note that, in the above examples, the returned references have no entitlements. 
+i.e: they are non-auth references.
+
+To get entitled references for struct/resource fields, they must be defined with [entitlement mappings](./access-control.md#entitlement-mappings).
+However, accessing a value at an index/key of an array/dictionary reference would always return a non-auth reference.
+
+### Index Assignment
+
+Assigning to an index of an array or a dictionary reference is an entitled-operation.
+In other words, the assignment operator for arrays/dictionaries would also have the `Mutate` and `Insert`
+[built-in entitlements](./access-control.md#built-in-mutability-entitlements).
+
+Think of assignment as a built-in function with `Mutate` and `Insert` entitlements. e.g:
+
+```cadence
+access(Mutate | Insert) set(keyOrIndex, value) { ... }
```

Comment:
Didn't we determine that this actually needs `(Mutate | (Insert && Remove))`, since it's not safe to do an indexed assignment with just a `Mutate` entitlement (https://github.com/onflow/cadence/pull/2664#pullrequestreview-1537718220)

Reply:
ah yes, I was mostly copying the stuff from the FLIP, had forgotten that this was changed during the implementation.

Reply:
updated in 8d8fd5c116e1d50f68e45f4a5b9c03543103c46e